src/Parser.js

```diff
@@ -543,47 +543,49 @@ class Parser {
             }, "math");
         } else {
             // Defer to parseFunction if it's not a function we handle
-            return this.parseFunction(start);
+            return this.parseGivenFunction(start);
         }
     }
 
     /**
      * Parses an entire function, including its base and all of its arguments.
-     * The base might either have been parsed already, in which case
-     * it is provided as an argument, or it's the next group in the input.
+     * It also handles the case where the parsed node is not a function.
      *
-     * @param {ParseFuncOrArgument=} baseGroup optional as described above
      * @return {?ParseNode}
      */
-    parseFunction(baseGroup) {
-        if (!baseGroup) {
-            baseGroup = this.parseGroup();
-        }
-
-        if (baseGroup) {
-            if (baseGroup.isFunction) {
-                const func = baseGroup.result;
-                const funcData = functions[func];
-                if (this.mode === "text" && !funcData.allowedInText) {
-                    throw new ParseError(
-                        "Can't use function '" + func + "' in text mode",
-                        baseGroup.token);
-                } else if (this.mode === "math" &&
-                    funcData.allowedInMath === false) {
-                    throw new ParseError(
-                        "Can't use function '" + func + "' in math mode",
-                        baseGroup.token);
-                }
+    parseFunction() {
+        const baseGroup = this.parseGroup();
+        return baseGroup ? this.parseGivenFunction(baseGroup) : null;
+    }
 
-                const args = this.parseArguments(func, funcData);
-                const token = baseGroup.token;
-                const result = this.callFunction(func, args, token);
-                return new ParseNode(result.type, result, this.mode);
-            } else {
-                return baseGroup.result;
+    /**
+     * Same as parseFunction(), except that the base is provided, guaranteeing a
+     * non-nullable result.
+     *
+     * @param {ParseFuncOrArgument} baseGroup
+     * @return {ParseNode}
+     */
+    parseGivenFunction(baseGroup) {
+        if (baseGroup.isFunction) {
```

Comment:
I feel like by storing `isFunction` on `ParseFuncOrArgument` it leads to unnecessary wrapping and unwrapping.  What if we create a function that given a `ParseNode` would tell us if it was a function?  With that, we should be able to get rid of `ParseFuncOrArgument` and just use `ParseNode` everywhere.  This is maybe something to consider for a future PR.

Reply:
Yes, I intend to do more changes within `Parser.js` itself in subsequent PRs while the `ParseNode` refactor waits on other PRs to complete. `ParseFuncOrArgument` will be replaced with a union of 3 structs per our discussion in #892. We can see whether it makes sense to have a function for an is-function check or simply use `foo.type === "fn"` (or `"$"` or `"arg"`) or some such.